[config] New wallet files handling

alexlyp opened the pull request:
This PR was due to an incoming feature for selecting alternative wallets instead of using the main decrediton app directory for wallet information storage.  

Now there is a sub-directory called `wallets`.  Within that directory will contain all the wallets each with their own directory.  For this beginning PR there is just "default-wallet."  If the user has an extisting wallet set up, this `default-wallet` directory will be made to match their existing wallets.  If no wallets exist, then this will be created and set up for them.

We have also changed the handling of various configuration settings.  There is a "global config" that will remain in the decrediton app directory and a wallet specific config.json that will be located in that wallets sub-directory.  Also each wallet set up has new dcr*.conf created for them on creation.  If the user would like to use advanced settings they may replace anything they'd like in these .confs and those will be used instead.

alexlyp commented:
I have tested this on mainnet/testnet and on a fresh install and from a legacy setup.

matheusd commented:
Test results. Unless otherwise noted, tests were on linux.

- Previously, running advanced_daemon=true and then skipping would connect to a previously running dcrd instance, but now it doesn't. I usually ran like that, but I understand that's actually a broken flow on my end. Just wanted to document it.

- Shuting down when using `yarn {start|dev}` to run seems to be broken (the dcrd/dcrwallet processes get shut down, but the Decrediton process stays up). 

- Migration of existing wallet (both mainnet and testnet) worked

- New wallet creation (both mainnet and testnet) worked

---

I'm a little undecided on the dir architecture. I think I'd prefer that each individual `wallet/` subdir store a single wallet+network combination, instead of each wallet being runnable either in testnet or mainnet mode. IMO, `wallets/wallet-foo/{mainnet|testnet}` means that the same seed for wallet-foo is being run over both testnet and mainnet (though I know that is not necessarily the case).

I would refactor to either each `wallets/wallet-name` store config for a single network **or** have `wallets/{mainnet|testnet}/wallet-name`, **but** I'm open for more discussion or input from other people on this, as I understand this would involve many more changes to the code.

alexlyp commented:
Right so my intention is to only have each wallet run on testnet or mainnet.  The subdir structure is mostly an issue with the way that dcrwallet creates the directories in the specified appdata directory.  

I do like the idea of have `wallets/{mainnet|testnet}/wallet-name`  let me investigate how involvled that would be too change, but shouldn't be too bad

alexlyp commented:
I believe I've updated it to use the recommended pathing.

Also, to confirm, yes I've removed the skip_daemon/wallet options.  To replicate you can just start the remote daemon and enter it's information in the advanced screen.

alexlyp commented:
Still investigating the shutdown oddity though.

alexlyp commented:
OK believe the shutdown error has been fixed as well.

matheusd commented on app/config.js:
```diff
@@ -163,23 +142,13 @@ export function getDcrdPath() {
-    certPath = path.join(os.homedir(), ".config", "decrediton", "rpc.cert");
-  }
+export function getWalletPath(testnet, walletPath) {
+  return path.join(appDataDirectory(), "wallets", testnet ? "testnet" : "mainnet", walletPath);
```
On a second review, this seems to depend. When called from the main node process, it gets a boolean, but when called from the react app it gets a string. Probably passing the `network` selector instead of `isTestnet` somewhere.

matheusd commented on app/main.development.js:
```diff
@@ -226,56 +264,50 @@ const installExtensions = async () => {
   }
   try {
-    dcrdPID = launchDCRD(appData);
+    dcrdPID = launchDCRD(walletPath, appData, testnet);
```
This is what is preventing the clean shutdown. `launchDCRD` is **not** returning the pid, but a `dcrdConfig` object, and then when shutting down, the `closeDcdr()` call doesn't actually close the daemon.

matheusd commented:
Ok, latest batch of tests:

- Wallet migration worked (testnet & mainnet)
- Shutdown worked

- Creation of new wallet (without existing to migrate) failed with a dcrwallet argument error (the ticketbuyer args are being sent as undefined)
- Importing a stakepool api failed (with that old stakey never goes away bug)
